bundles/org.openhab.binding.monopriceaudio/src/main/java/org/openhab/binding/monopriceaudio/internal/communication/MonopriceAudioConnector.java

```diff
@@ -0,0 +1,264 @@
+     */
+    protected void setConnected(boolean connected) {
+        this.connected = connected;
+    }
+
+    /**
+     * Set the thread that handles the feedback messages
+     *
+     * @param readerThread the thread
+     */
+    protected void setReaderThread(Thread readerThread) {
+        this.readerThread = readerThread;
+    }
+
+    /**
+     * Open the connection with the MonopriceAudio device
+     *
+     * @throws MonopriceAudioException - In case of any problem
+     */
+    public abstract void open() throws MonopriceAudioException;
+
+    /**
+     * Close the connection with the MonopriceAudio device
+     */
+    public abstract void close();
+
+    /**
+     * Stop the thread that handles the feedback messages and close the opened input and output streams
+     */
+    protected void cleanup() {
+        Thread readerThread = this.readerThread;
+        OutputStream dataOut = this.dataOut;
+        if (dataOut != null) {
+            try {
+                dataOut.close();
+            } catch (IOException e) {
+                logger.debug("Error closing dataOut: {}", e.getMessage());
+            }
+            this.dataOut = null;
+        }
+        InputStream dataIn = this.dataIn;
+        if (dataIn != null) {
+            try {
+                dataIn.close();
+            } catch (IOException e) {
+                logger.debug("Error closing dataIn: {}", e.getMessage());
+            }
+            this.dataIn = null;
+        }
+        if (readerThread != null) {
+            readerThread.interrupt();
+            try {
+                readerThread.join(3000);
+            } catch (InterruptedException e) {
+                logger.warn("Error joining readerThread: {}", e.getMessage());
+            }
+            this.readerThread = null;
+        }
+    }
+
+    /**
+     * Reads some number of bytes from the input stream and stores them into the buffer array b. The number of bytes
+     * actually read is returned as an integer.
+     *
+     * @param dataBuffer the buffer into which the data is read.
+     *
+     * @return the total number of bytes read into the buffer, or -1 if there is no more data because the end of the
+     *         stream has been reached.
+     *
+     * @throws MonopriceAudioException - If the input stream is null, if the first byte cannot be read for any reason
+     *             other than the end of the file, if the input stream has been closed, or if some other I/O error
+     *             occurs.
+     */
+    protected int readInput(byte[] dataBuffer) throws MonopriceAudioException {
+        InputStream dataIn = this.dataIn;
+        if (dataIn == null) {
+            throw new MonopriceAudioException("readInput failed: input stream is null");
+        }
+        try {
+            return dataIn.read(dataBuffer);
+        } catch (IOException e) {
+            throw new MonopriceAudioException("readInput failed: " + e.getMessage());
```

Comment:
You should include the cause exception in your MonopriceAudioException.
```suggestion
            throw new MonopriceAudioException("readInput failed: " + e.getMessage(), e);
```

Reply:
done for all